icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitPreferences.java

```diff
@@ -56,14 +66,46 @@ private static String[] getAllUsages(String usage) {
         return result.toArray(new String[0]);
     }
 
-    public UnitPreference[] getPreferencesFor(String category, String usage, String region) {
+    public UnitPreference[] getPreferencesFor(String category, String usage, ULocale locale, UnitsData data) {
+        // TODO: remove this condition when all the categories are allowed.
+        if (category.equals("temperature")) {
+            String localeUnit = locale.getKeywordValue("measurement-unit");
+            String localeUnitCategory;
+            try {
+                localeUnitCategory = localeUnit == null ? null : data.getCategory(MeasureUnitImpl.forIdentifier(localeUnit));
+            } catch (Exception e){
+                localeUnitCategory = null;
+            }
+
+            if (category.equals(localeUnitCategory)) {
+                UnitPreference[] preferences = {new UnitPreference(localeUnit, null, null)};
+                return preferences;
+            }
+        }
+
+        String region = locale.getCountry();
+
+        String localeSystem = locale.getKeywordValue("measure");
+        boolean isLocaleSystem = false;
+        if (measurementSystem.containsKey(localeSystem)){
+            isLocaleSystem = true;
+            region = measurementSystem.get(localeSystem);
+        }
+
+        String localeRegion = locale.getKeywordValue("rg");
+        if (!isLocaleSystem && localeRegion != null) {
+            // we need to check the region and the capitalization.
+            region = localeRegion.replaceAll("zzzz", "").toUpperCase();
```

Comment:
The following are all potentially valid:

- `gbzzzz`
- `150zzzz`
- `usca`

You want to select the first 2 chars if the first char is a letter, and the first 3 chars if the first char is a number.

Reply:
In C++, you can make a stack string: `UChar region[] = {0, 0, 0, 0};`

Reply:
That code is inefficient. It is simpler to just extract the first two characters, eg 

localeRegion.substring(0,2).toUpperCase(Locale.ROOT);

Also, the upper/lowercase methods have to have an explicit locale. Otherwise they can screwup if the default locale is, say, Turkish.

Reply:
You would also want to check that localeRegion.size() > 2 before substringing.

Reply:
Done